File: 思酷在线RAS加密/skzx.py
# http://study.study2win.net/wx/register.html  注册地址
# http://study.study2win.net/wx/login.html  登录地址

from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5
import time, math, random, json, base64, requests


def get_json(a):
    nonce = ""
    for i in range(6):
        nonce += str(math.floor(random.random() * 10))
    nonce = str(int(time.time()*1000)) + nonce
    data = {
        "text":a,
        "timestamp":int(time.time()*1000),
        "nonce":nonce
    }
    return json.dumps(data)


def rsa_encrypt(a):
    key = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEApQUOaXuFw6ecagtlJQ6a7lnxTSLtTFVQ/Sf0iA/P45zX44AshFFG3iGzcNiTPTepGMGuoLm1LYzdT1s8NrVQ+qgmSAOSN9x//hcJdjSdh2vXVwDDANIryw0fxmsdRbNjPsUvL4j8vulRulPpUhy2v5/LyxyDWOzOjG2IftZxbFXsc2yLohpC1h1vUADAF50D1eUm2ilJXNCNWrofjwFiK2J1Q3h6TUW7ZJhwtFCm3DoUIAK2WU7XOAt/rMJopULbU1ThL+UdUydDEzLBEDIjD8SS3E9NcILatGuc7xr8re+KPLqdiL7Nmmvbb9toJq39fnbYeNrBUlKSJPfRo8HG7QIDAQAB"
    publickey = """-----BEGIN PUBLIC KEY-----
    {key}
    -----END PUBLIC KEY-----""".format(key=key)

    rsakey = RSA.importKey(publickey)
    cipher = PKCS1_v1_5.new(rsakey)
    cipher_text = cipher.encrypt(get_json(a).encode("utf-8"))
    return '_SCHO_ENC_:V1:RSA:ZSXYPROD0001:' + str(base64.b64encode(cipher_text), encoding="utf-8")


url = "https://study.study2win.net/front/login/unp"
data = {
    "orgCode":"SAAS",
    "password":rsa_encrypt("111111"),
    "svcCode":"client:009",
    "username":rsa_encrypt("15800000000"),
}
# print(data)
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36',
    'Referer': 'http://study.study2win.net/wx/login.html',
    'Origin': 'https://study.study2win.net',
    'Content-Type': 'application/json',
}
res = requests.post(url, headers=headers, data=json.dumps(data))
print(res.json())

File: 09-思酷在线(RSA)/skzx.js
//RAS2048密码加密
var RAS = {
    rasEncrypt: function (password) {
        var nonce = "";  //随机8号数以上
        for (var i = 0; i < 6; i++) {  //6位随机数，用以加在时间戳后面。
            nonce += Math.floor(Math.random() * 10);
        }
        nonce = new Date().getTime() + nonce;  //时间戳，用来生成订单号。

        //转换成需要的格式
        var data = {
            text: password,
            timestamp: new Date().getTime(),
            nonce: nonce
        };

        // var host = window.location.host;
        //公钥：生产环境，包括：标品、中特独立部署、NBCB独立部署或其他独立部署；
        var Key = 'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEApQUOaXuFw6ecagtlJQ6a7lnxTSLtTFVQ/Sf0iA/P45zX44AshFFG3iGzcNiTPTepGMGuoLm1LYzdT1s8NrVQ+qgmSAOSN9x//hcJdjSdh2vXVwDDANIryw0fxmsdRbNjPsUvL4j8vulRulPpUhy2v5/LyxyDWOzOjG2IftZxbFXsc2yLohpC1h1vUADAF50D1eUm2ilJXNCNWrofjwFiK2J1Q3h6TUW7ZJhwtFCm3DoUIAK2WU7XOAt/rMJopULbU1ThL+UdUydDEzLBEDIjD8SS3E9NcILatGuc7xr8re+KPLqdiL7Nmmvbb9toJq39fnbYeNrBUlKSJPfRo8HG7QIDAQAB';
        var Code = '_SCHO_ENC_:V1:RSA:ZSXYPROD0001:';
        //公钥：测试环境，包括：本地测试，BETA测试
        // var Key = 'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA10AjU2HIhIb9hnjsRkHvPUVT91pl9fR9VKn/F/JbwrNlDZQOnd0AXpIM9tiHX2hK2aoV0MS4pTPWpAyC722fdwcupvDquRSlfU7TRI6mRPXo9ALHEUYIA2Bnpt0lU8VcP61EVOdEqAdtA1law/6Z9O4c1nHaDBblx3R9Sr7Lw3KJj6P2pRM/eNCrMDxw2PRf886UWSbJLKlvx0kxoox7LlAInToUqU1ofWNf0FlF+A6kd1wZhil1Iha9NS8z7UfMx92jxh9RtGWFK0gredeKFR1S7lAKjnW1bUzjrWvPmiEl4UJsQoS7krDN6skb8SLwga4QYUU3ua8GCRZBPZJ4QQIDAQAB';
        // var Code = '_SCHO_ENC_:V1:RSA:ZSXYDEV0001:';

        //-----BEGIN PUBLIC KEY-----和-----END PUBLIC KEY-----，是占位符，替换公钥时要保留
        var rsaPublicKey = KEYUTIL.getKey("-----BEGIN PUBLIC KEY-----" + Key + "-----END PUBLIC KEY-----");

        //加密的明文
        var hexStr = KJUR.crypto.Cipher.encrypt(JSON.stringify(data), rsaPublicKey, "RSA");

        //输出的密文
        return Code + RAS.hexToBase64(hexStr);
    },
    // Hex to Base64
    hexToBase64: function (str) {
        return String.fromCharCode.apply(null,
            str.replace(/\r|\n/g, "").replace(/([\da-fA-F]{2}) ?/g, "0x$1 ").replace(/ +$/, "").split(" "));
    }
};

